File: Data Structures & Algorithms/merge-intervals/submission-4.ts
class Solution {
    /**
     * @param {number[][]} intervals
     * @return {number[][]}
     */
    merge(intervals: number[][]): number[][] {
        intervals.sort((a, b) => a[0] - b[0]);

        const result = [intervals[0]];
        for (let i = 1; i < intervals.length; i++) {
            const [currentStart, currentEnd] = result[result.length-1];
            const [nextStart, nextEnd] = intervals[i];

            if (currentEnd >= nextStart) {
                // merge
                result[result.length-1] = [currentStart, Math.max(currentEnd, nextEnd)];
            } else {
                // new interval
                result.push(intervals[i]);
            }
        }

        return result;
    }
}

File: Data Structures & Algorithms/non-overlapping-intervals/submission-4.java
class Solution {
    public int eraseOverlapIntervals(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[1], b[1]));
        //System.out.println(Arrays.deepToString(intervals));

        int currentEnd = intervals[0][1];
        int result = 0;

        for (int i = 1; i < intervals.length; i++) {
            int nextIntervalBeginning = intervals[i][0];
            int nextIntevalEnd = intervals[i][1];
            if (nextIntervalBeginning < currentEnd) {
                result++;
            } else {
                currentEnd = nextIntevalEnd;
            }
        }

        return result;
    }
}
